### PR TITLE
Add a `MessagePort` to the `AudioWorkletGlobalScope`

### LINKED ISSUE #2456: How to postMessage to AudioWorkletGlobalScope without creating a dummy worklet processor + node?
When implementing WebAssembly integration with Web Audio AudioWorklets with Shared Memory, one issue is that AudioWorkletGlobalScope does not seem to have a postMessage/onMessage type of functionality.

As a result, in order to share the WebAssembly application Module and Memory objects with the AudioWorkletGlobalScope, one needs to create a dummy processor and instantiate it on the main thread in order to be able to send these bits of data over.

Unless I have missed something, this looks like a design shortcoming of the spec - and results in one having to abuse the spec in order to achieve the necessary goal.

You can see examples of this abuse in two different implementations of Shared Memory Wasm <-> AudioWorklet integrations:
 - https://github.com/juj/emscripten/compare/wasm_workers...juj:audio_worklets?expand=1#diff-28e06df107fc96633117cec6045e3b962716e05474542608b47129554ecf263cR120 and
 - https://github.com/emscripten-core/emscripten/pull/12502/files#diff-31ce84d9e9748b123befe225a55c67718b58011f9b0a099c821e5c5f0019b1faR547

I wonder if I've missed a critical piece of the spec and there is a simpler way to do this - or did this use case fall through in the spec?

If so, would it be possible to add a postMessage()/onMessage() functionality to AudioWorkletGlobalScope? That would simplify the creation and management of shared memory Wasm Audio Worklets.

### PR DESCRIPTION
**Pull request by padenot**
This fixes #2456.

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Sep 19, 2022, 10:13 PM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [CSS Spec Preprocessor](https://api.csswg.org/bikeshed/) - CSS Spec Preprocessor is the web service used to build Bikeshed specs.

:link: [Related URL](https://api.csswg.org/bikeshed/?url=https%3A%2F%2Fraw.githubusercontent.com%2FWebAudio%2Fweb-audio-api%2Fadf9c503ed2e736af16f36c6ecbdfbbd065ecb38%2Findex.bs&force=1&md-warning=not%20ready)

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 [no address given] to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
<hr>
<address>Apache/2.4.10 (Debian) Server at api.csswg.org Port 443</address>
</body></html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20WebAudio/web-audio-api%232507.)._
</details>

### CODE REVIEW

**Comment by padenot** [comment]
This does not have the markup at the moment.

File: index.bs [hunk 1]
```diff
@@ -1596,11 +1596,31 @@ Constructors</h4>
 
 			2. Set a {{[[rendering thread state]]}} to <code>suspended</code> on the {{AudioContext}}.
 
-			3. Let <dfn attribute for="AudioContext">[[pending resume promises]]</dfn> be a
+			3. Let <var>messageChannel</var> be a new {{MessageChannel}}.
```
Comment (hoch) [inline index.bs:1599]: I am okay with making all these numbered list to `1.`.

File: index.bs [hunk 2]
```diff
@@ -1596,11 +1596,31 @@ Constructors</h4>
...
+
+			4. Let <var>controlSidePort</var> be the value of
```
Comment (hoch) [inline index.bs:1601]: Can we use `|controlSidePort|` notation instead of using `<var>`?

File: index.bs [hunk 3]
```diff
@@ -10347,10 +10421,19 @@ Attributes</h5>
 	: <dfn>sampleRate</dfn>
 	::
 		The sample rate of the associated {{BaseAudioContext}}.
-</dl>
 
-<h5 id="AudioWorkletGlobalScope-methods">
-Methods</h5>
```
Comment (hoch) [inline index.bs:10353]: I think these are accidentally removed.

**Review by hoch — COMMENTED** [review]
Looks great, but a couple of nits.

File: index.bs [hunk 4]
```diff
@@ -1594,19 +1594,39 @@ Constructors</h4>
 
 			1. Set a {{[[control thread state]]}} to <code>suspended</code> on the {{AudioContext}}.
 
-			2. Set a {{[[rendering thread state]]}} to <code>suspended</code> on the {{AudioContext}}.
+			1. Set a {{[[rendering thread state]]}} to <code>suspended</code> on the {{AudioContext}}.
 
-			3. Let <dfn attribute for="AudioContext">[[pending resume promises]]</dfn> be a
+			1. Let <var>messageChannel</var> be a new {{MessageChannel}}.
```
Comment (hoch) [inline index.bs:1599]: Same here. Using `|messageChannel|` should be simpler and shorter.

**Review by hoch — APPROVED** [review]
LGTM with nits

File: index.bs (hunk 4 above)
Comment (hoch) [reply index.bs:1599]: I meant using the pipe to denote algorithm variables.

**Comment by guest271314** [comment]
So this exposes two (2) `MesagePort`s in `AudioWorkletGlobalScope`? The `port` in `AudioWorkletProcessor` construction does not get changed or removed?

File: index.bs (hunk 4 above)
Comment (padenot) [reply index.bs]: Fixed, and added the necessary markup.

File: index.bs [hunk 5]
```diff
@@ -10175,11 +10318,46 @@ Dictionary {{WaveShaperOptions}} Members</h5>
...
+
+	<h4 id="AudioWorklet-attributes">Attributes</h4>
+
+	<dl dfn-type=attribute dfn-for="AudioWorklet">
+		: <dfn>port</dfn>
+		::
+			A {{MessagePort}} connected to the port on the
+			{{AudioWorkletGlobalScope}}.
+
+			Note: Authors that register a event listener on the
```
Comment (hoch) [inline index.bs:10352]: typo: a -> an

File: index.bs [hunk 6]
```diff
@@ -10313,18 +10491,44 @@ Note: An {{AudioWorkletGlobalScope}} is associated with a single
...
+Add a MessagePort to the AudioWorkletGlobalScope
+<div class="amendment-buttons">
+</div>
+<del cite=#2456>
+<xmp class="idl extract" data-no-idl>
+	callback AudioWorkletProcessorConstructor = AudioWorkletProcessor (object options);
+
+	[Global=(Worklet, AudioWorklet), Exposed=AudioWorklet]
+	interface AudioWorkletGlobalScope : WorkletGlobalScope {
+	undefined registerProcessor (DOMString name,
```
Comment (hoch) [inline index.bs:10508]: Perhaps it needs proper indentations?

File: index.bs [hunk 7]
```diff
@@ -10313,18 +10491,44 @@ Note: An {{AudioWorkletGlobalScope}} is associated with a single
...
+</del>
+<ins cite=#2456>
 <pre class="idl">
 callback AudioWorkletProcessorConstructor = AudioWorkletProcessor (object options);
 
 [Global=(Worklet, AudioWorklet), Exposed=AudioWorklet]
 interface AudioWorkletGlobalScope : WorkletGlobalScope {
 	undefined registerProcessor (DOMString name,
-	                             AudioWorkletProcessorConstructor processorCtor);
+								 AudioWorkletProcessorConstructor processorCtor);
```
Comment (hoch) [inline index.bs:10523]: Does this show up correctly after bikeshed processing?

File: index.bs [hunk 8]
```diff
@@ -10347,6 +10551,28 @@ Attributes</h5>
...
+	</span>
+	Add a MessagePort to the AudioWorkletGlobalScope
+	<div class="amendment-buttons">
+	</div>
+	<ins cite=#2456>
+		: <dfn>port</dfn>
+		::
+			A {{MessagePort}} connected to the port on the {{AudioWorklet}}.
+
+			Note: Authors that register a event listener on the <code>"message"</code>
```
Comment (hoch) [inline index.bs:10568]: a -> an

**Review by hoch — APPROVED** [review]
A couple of typos; let's fix them and merge.

File: index.bs (hunk 7 above)
Comment (padenot) [reply index.bs:10523]: It does yeah.

**Pull request merged**